crates/proof-of-sql/src/base/bit/bit_distribution.rs

```diff
@@ -1,70 +1,87 @@
 }
 
 impl BitDistribution {
     pub fn new<S: Scalar, T: Into<S> + Clone>(data: &[T]) -> Self {
-        if data.is_empty() {
-            return Self {
-                or_all: [0; 4],
-                vary_mask: [0; 4],
-            };
-        }
-        let mut or_all = make_abs_bit_mask(data[0].clone().into());
-        let mut vary_mask = [0; 4];
-        for x in data.iter().skip(1) {
-            let mask = make_abs_bit_mask((*x).clone().into());
-            for i in 0..4 {
-                vary_mask[i] |= or_all[i] ^ mask[i];
-                or_all[i] |= mask[i];
-            }
+        let bit_masks = data.iter().cloned().map(Into::<S>::into).map(make_bit_mask);
+        let (sign_mask, inverse_sign_mask) =
+            bit_masks
+                .clone()
+                .fold((U256::MAX, U256::MAX), |acc, bit_mask| {
+                    let bit_mask = if is_bit_mask_negative_representation(bit_mask) {
+                        bit_mask ^ (U256::MAX >> 1)
+                    } else {
+                        bit_mask
+                    };
+                    (acc.0 & bit_mask, acc.1 & !bit_mask)
+                });
+        let vary_mask_bit = U256::from(
+            !bit_masks
+                .map(is_bit_mask_negative_representation)
+                .all_equal(),
+        ) << 255;
+        let vary_mask: U256 = !(sign_mask | inverse_sign_mask) | vary_mask_bit;
+
+        Self {
+            leading_bit_mask: sign_mask.into(),
+            vary_mask: vary_mask.into(),
         }
-        Self { or_all, vary_mask }
     }
 
-    pub fn num_varying_bits(&self) -> usize {
-        let mut res = 0_usize;
-        for xi in &self.vary_mask {
-            res += xi.count_ones() as usize;
-        }
-        res
+    pub fn vary_mask(&self) -> U256 {
+        U256::from(self.vary_mask)
+    }
+
+    /// Identifies all columns that are the identical to the lead column.
+    pub fn leading_bit_mask(&self) -> U256 {
+        U256::from(self.leading_bit_mask) | (U256::ONE << 255)
+    }
+
+    /// Identifies all columns that are the identical to the inverse of the lead column.
+    pub fn leading_bit_inverse_mask(&self) -> U256 {
+        (!self.vary_mask() ^ self.leading_bit_mask()) & (U256::MAX >> 1)
     }
 
-    pub fn has_varying_sign_bit(&self) -> bool {
-        self.vary_mask[3] & (1 << 63) != 0
+    /// # Panics
+    ///
+    /// Panics if conversion from `ExpType` to `usize` fails
+    pub fn num_varying_bits(&self) -> usize {
+        self.vary_mask().count_ones() as usize
     }
 
-    #[allow(clippy::missing_panics_doc)]
-    pub fn sign_bit(&self) -> bool {
-        assert!(!self.has_varying_sign_bit());
-        self.or_all[3] & (1 << 63) != 0
+    /// # Panics
+    ///
+    /// Panics if lead bit varies but `bit_evals` is empty
+    pub fn leading_bit_eval<S: ScalarExt>(&self, bit_evals: &[S], one_eval: S) -> S {
+        if U256::from(self.vary_mask) & (U256::ONE << 255) != U256::ZERO {
```

Comment:
NIT:
I just realized this is supported. `shl` is const.
```suggestion
        if U256::from(self.vary_mask) & 256::ONE.shl(255) != U256::ZERO {
```